### ISSUE TITLE
Incorrect Button and Link Keyboard Behavior 

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#740]
The MathShare codebase utilizes many custom implementations of buttons and links.  These are marked up as `<div>` or `<span>` elements, and then augmented with accessibility properties and event handlers.  In other cases, a native `<button>` is used, but given a role of “link” and a custom `onKeyPress` event binding. 
 
A user will expect to be able to activate buttons and links using the space bar, Enter and no other keys.  In this case, a catch-all `onKeyPress` handler is connected to the controls, meaning that they can be activated using any key on the keyboard.  Many users will find this to be extremely confusing.

#### Suggested Solution 
 
These controls do not seem to offer any perceivable benefits over their native counterparts.  We recommend replacing them with native `<a>` and `<button>` elements, removing any `tabindex` attributes and custom keyboard event handlers along the way. 
 
Hyperlinks should be used, implemented as `<a>` elements, whenever activating a control would cause a full page reload on a more traditional, server-side website.  This includes cases when:

* the route in the URL changes, and along with it the text of the browser’s address bar;
* an entry is created in the browser history;
* there is a complete change in context (such as moving from the Landing Page to the application proper); or
* some content will open in a new window. 
 
By contrast, `<button>` elements should be used for controls which trigger a same-page action.  Examples include opening a modal, toggling the display of content or reformatting some text. 
 
If native `<a>` and `<button>` elements are correctly used in the appropriate situations, `role` attributes will not need to be applied to them.  There is one important caveat — screen readers will not recognize `<a>` elements without a `href` attribute as links.  You can get around this by correctly using the `href` attribute whenever possible, and applying a `href` attribute with a value of `"javascript:void(0)"` otherwise.  For consistency, you may consider using the [`<Link>` component from React Router](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/Link.md), which reports to generate accessible, semantic and valid anchor tags. 
 
Where native elements are not possible, an `onKeyPress` handler should be implemented which only responds to Enter and the space bar.  However, we would stress that custom buttons and links should only be necessary in truly exceptional circumstances.

**Comment by benetechMatt** [comment]
listed sets on the main page use <button> for "new" and example. pre-made sets use <a>. Both use Sina's suggested but will this result in different behavior or expectations for the user?

Edit and delete buttons on problem cards are span.

"finish" button at bottom of a problem should have role="link" because users are taken to a new page

"continue without sign in" button missing role=link
Microsoft sign in is a button. no issues
Google sign in appears to only be divs, no button or <a>

homepage:
"go to app" button missing role=link
"sign in" button missing role=link


keypress:
pressing spacebar on the pre-made sets scrolls to bottom instead of opening it
all keys activate assign to microsoft teams button when editing your own problem set. works correctly on pre-made sets/student view

**Comment by rupeshparab** [comment]
> listed sets on the main page use < button > for "new" and example. pre-made sets use < a >. Both use Sina's suggested but will this result in different behavior or expectations for the user?

I am following a pattern where if the next page's (after clicking on a button) is determinable before clicking on the link, I am using <a> and specifying the link in href
In cases where the link is not deterministic before the click (in case of new problemset, link is generated after click), I am using a button

got the 2 cases of new and example, I have changed example to <a> as its link can be determined similar to pre-made sets

I have also fixed rest of the mentioned issues

**Comment by jscholes** [comment]
@rupeshparab - there are some outstanding cases of buttons having a "link" role (and maybe vice versa).  For example, the control to add a new problem set should have its role of "link" removed, because it opens a modal.  That's a same-page action and should always use a button.

I will look through as many of the controls as I can, and attempt to come up with a list of other examples.  However, I did want to ask about commit https://github.com/benetech/MathShare/commit/9e986c313c86b74fecd5b1b720fefdd7e5ea8a9d.

Each problem set is now (correctly) marked up as a hyperlink with an href attribute.  So are the `onClick` and `onKeyPress` event handlers still necessary?  The browser will handle mouse and keyboard behavior, and navigate the user to the right place.

**Comment by rupeshparab** [comment]
@jscholes 
> Each problem set is now (correctly) marked up as a hyperlink with an href attribute. So are the `onClick` and `onKeyPress` event handlers still necessary? The browser will handle mouse and keyboard behavior, and navigate the user to the right place.

The onClick and onKeyPress are being used to send Google Analytics events when a user clicks on a link

**Comment by jscholes** [comment]
@rupeshparab - got it.

**Comment by johnhbenetech** [comment]
so to summarize it seems like the control to add a new problem set should have its role of "link" removed, because it opens a modal.

Other potential issues:
* continue without sign in on sign in modal
* Sign in on landing  page in upper right
* I don't know where this is as i cant see the role on the finish button in the frontend: https://github.com/benetech/MathShare/blob/782625f44d28b6560867da047e5640c8aab1efd2/src/components/Editor/components/MySteps/components/MyStepsList/index.js#L172

**Comment by rupeshparab** [comment]
I have removed the incorrect roles of link from buttons and wherever possible I have changed button to anchor

**Comment by jscholes** [comment]
@rupeshparab - when you say "where possible", could you indicate the cases in which it wasn't?

**Comment by rupeshparab** [comment]
@jscholes 
While searching for 'role="link"', I was able to find 5 instances, of these I changed 2 buttons to anchors, these are the other 3 buttons where I have just removed the role link:
Finish Button in solution view: https://github.com/benetech/MathShare/commit/9c8d762c2390a6831a61b3905adbf3790866ba19#diff-bee05272ed39cebf379d4bfdc36409bcL172
New problem set button:
https://github.com/benetech/MathShare/commit/9c8d762c2390a6831a61b3905adbf3790866ba19#diff-3e5a4ba04ec71593d83ba914ec69cc92R190
Go Back button:
https://github.com/benetech/MathShare/commit/9c8d762c2390a6831a61b3905adbf3790866ba19#diff-572d509dd96156866bd53afaca7de546L109

In all of these cases keeping the button tag seemed more appropriate as there isn't any fixed link which they point to

**Comment by jscholes** [comment]
@rupeshparab - I agree in all cases but the last, because the "Continue without signing in" button does cause navigation.  Is it possible to pull a URL out of the history and set that as the `href` on a link, rather than using `history.goBack`?

Otherwise, an `<a>` with an `href` of `javascript:void(0)` would also work, and then do the history nav in an `onClick`.

**Comment by rupeshparab** [comment]
@jscholes 
If we use the previous page's url in anchor it will add the previous link as a new link to the history and won't pop the current state, I think it will be best to rely on the goBack function to pop the current link

I would prefer if we could avoid the 'javascript:void(0)' approach where having a link isn't necessary like the skip link. When we had used 'javascript:void(0)' in href for skip link, I was getting this warning in dev environment: 'react-dom.development.js:558 Warning: A future version of React will block javascript: URLs as a security precaution. Use event handlers instead if you can. If you need to generate unsafe HTML try using dangerouslySetInnerHTML instead. React was passed "javascript:void(0)"'

**Comment by jscholes** [comment]
@rupeshparab - that's fair.  It's implemented as a `<button>` so it already has the necessary keyboard behaviour.  `role="link"` it is!

**Comment by rupeshparab** [comment]
I have added the role link to the button

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/4f6e30c606111c7ce5004d57e5319e86b02178b4.
